tactic/where.lean

```diff
@@ -0,0 +1,226 @@
+import tactic.basic
+
+open lean.parser tactic
+
+namespace where
+
+meta def mk_flag (let_var : option name := none) : lean.parser (name × ℕ) :=
+do n ← mk_user_fresh_name,
+   emit_code_here $ match let_var with
+   | none   := sformat!"def {n} := ()"
+   | some v := sformat!"def {n} := let {v} := {v} in ()"
+   end,
+   nfull ← resolve_constant n,
+   return (nfull, n.components.length)
+
+meta def get_namespace_core : name × ℕ → name
+| (nfull, l) := nfull.get_nth_prefix l
+
+meta def resolve_var : list name → ℕ → expr
+| [] _ := default expr
+| (n :: rest) 0 := expr.const n []
+| (v :: rest) (n + 1) := resolve_var rest n
+
+meta def resolve_vars_aux : list name → expr → expr
+| head (expr.var n) := resolve_var head n
+| head (expr.app f a) := expr.app (resolve_vars_aux head f) (resolve_vars_aux head a)
+| head (expr.macro m e) := expr.macro m $ e.map (resolve_vars_aux head)
+| head (expr.mvar n m e) := expr.mvar n m $ resolve_vars_aux head e
+| head (expr.pi n bi t v) :=
+  expr.pi n bi (resolve_vars_aux head t) (resolve_vars_aux (n :: head) v)
+| head (expr.lam n bi t v) :=
+  expr.lam n bi (resolve_vars_aux head t) (resolve_vars_aux (n :: head) v)
+| head e := e
+
+meta def resolve_vars : expr → expr :=
+resolve_vars_aux []
+
+meta def strip_pi_binders_aux : expr → list (name × binder_info × expr)
+| (expr.pi n bi t b) := (n, bi, t) :: strip_pi_binders_aux b
+| _ := []
+
+meta def strip_pi_binders : expr → list (name × binder_info × expr) :=
+strip_pi_binders_aux ∘ resolve_vars
+
+meta def get_def_variables (n : name) : tactic (list (name × binder_info × expr)) :=
+(strip_pi_binders ∘ declaration.type) <$> get_decl n
+
+meta def get_includes_core (flag : name) : tactic (list (name × binder_info × expr)) :=
+get_def_variables flag
+
+meta def binder_brackets : binder_info → string × string
+| binder_info.implicit        := ("{", "}")
+| binder_info.strict_implicit := ("{", "}")
+| binder_info.inst_implicit   := ("[", "]")
+| _                           := ("(", ")")
+
+meta def binder_priority : binder_info → ℕ
+| binder_info.implicit := 1
+| binder_info.strict_implicit := 2
+| binder_info.default := 3
+| binder_info.inst_implicit := 4
+| binder_info.aux_decl := 5
+
+meta def binder_less_important (u v : binder_info) : bool :=
+(binder_priority u) < (binder_priority v)
+
+meta def is_in_namespace_nonsynthetic (ns n : name) : bool :=
+ns.is_prefix_of n ∧ ¬(ns.append `user__).is_prefix_of n
+
+meta def get_all_in_namespace (ns : name) : tactic (list name) :=
+do e ← get_env,
+   return $ e.fold [] $ λ d l,
+     if is_in_namespace_nonsynthetic ns d.to_name then d.to_name :: l else l
+
+meta def erase_duplicates {α : Type} [decidable_eq α] : list α → list α
+| [] := []
+| (x :: t) := (x :: erase_duplicates (t.filter $ λ a, a ≠ x))
```

Comment:
This should be `erase_dup` now that cyclic import hell is avoided

Reply:
thanks mario